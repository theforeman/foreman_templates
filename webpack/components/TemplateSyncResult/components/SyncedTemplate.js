import React from 'react';
import { ListView, Grid, Icon, OverlayTrigger, Tooltip } from 'patternfly-react';
import classNames from 'classnames';
import { pick, mergeWith, isEmpty } from 'lodash';

const StringInfoItem = ({ template, attr, tooltipText, translate = false }) => {
    return (<InfoItem itemId={itemIteratorId(template, attr)}
                      tooltipText={tooltipText}>
              <strong>{ translate ? __(template[attr]) : template[attr] }</strong>
            </InfoItem>);
};

const IconInfoItem = ({ template, attr, cssClassNames, tooltipText }) => {
  return (<InfoItem itemId={itemIteratorId(template, attr)}
                    tooltipText={tooltipText}>
            <span className={cssClassNames} />
          </InfoItem>);
};

const EmptyInfoItem = (template, attr) => (
  <InfoItem itemId={itemIteratorId(template, attr)} />
);

const InfoItem = ({ itemId, children, tooltipText }) => {
  const overlay = (
    <OverlayTrigger overlay={tooltipText ? (<Tooltip id={itemId}>{ tooltipText }</Tooltip>) : ''}
                    placement="top"
                    trigger={['hover', 'focus']}
                    rootClose={false}
                    >
        { children }
      </OverlayTrigger>
    )
  return (
    <ListView.InfoItem key={itemId} className='additional-info-wide'>
      { tooltipText ? overlay : children }
    </ListView.InfoItem>
  );
}

const itemIteratorId = (template, attr) =>
  `${template.name}-${attr}`;

const SyncedTemplate = props => {
  const { template } = props;

  const additionalInfo = (template) => {
    const infoAttrs = ['locked', 'snippet', 'class_name', 'kind'];

    return infoAttrs.map((attr) => {
      const key = itemIteratorId(template, attr)

      if (!template[attr]) {
        return (<EmptyInfoItem template={template} attr={attr} key={key}/>);
      }

      switch(attr) {
        case 'locked':
          return (<IconInfoItem template={template}
                                attr={attr}
                                cssClassNames={'glyphicon glyphicon-lock'}
                                tooltipText='Locked'
                                key={key} />)
        case 'snippet':
          return (<IconInfoItem template={template}
                                attr={attr}
                                cssClassNames={'glyphicon glyphicon-scissors'}
                                tooltipText={'Snippet'}
                                key={key} />);
        case 'class_name':
          return (<StringInfoItem template={template}
                                  attr={attr}
                                  translate={true}
                                  key={key} />);
        case 'kind':
          return (<StringInfoItem template={template}
                                  attr={attr}
                                  key={key} />);
      }
    });
  };

  const aggregatedErrors = template => {
    const err = { ...template.errors } || {};
    if (template.additional_errors) {
      err.additional = template.additional_errors
    }

    return err;
  }

  const itemLeftContentIcon = (template) => {
    const iconName =  isEmpty(aggregatedErrors(template)) ? 'ok' : 'error-circle-o';
    return (<Icon name={iconName} size="sm" type={'pf'} />);
  }

  const formatError = (key, value) => {
    const omitKeys = ['base', 'additional'];
    if (omitKeys.reduce((memo, item) => (memo || key === item), false)) {
      return value;
    }

    return `${key}: ${value}`;
  }

  const templateErrors = (template) => {
    if (Object.keys(aggregatedErrors(template)).length !== 0) {
      const res = Object.keys(aggregatedErrors(template)).map((key) => {
        return (<li key={itemIteratorId(template, key)}>{ formatError(key, aggregatedErrors(template)[key]) }</li>)
      });
      return (<ul>{ res }</ul>);
    }
    return (<span>There were no errors.</span>);
  }

  return (
      <ListView.Item
        key={template.id}
        heading={template.name}
        additionalInfo={additionalInfo(template)}
        className={'listViewItem--listItemVariants'}
        leftContent={itemLeftContentIcon(template)}
        hideCloseIcon
        stacked
      >
        { templateErrors(template) }
    </ListView.Item>
  );
}

export default SyncedTemplate;
