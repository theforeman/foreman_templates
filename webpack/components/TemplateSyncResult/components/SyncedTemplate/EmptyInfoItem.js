import React from 'react';
import PropTypes from 'prop-types';

import InfoItem from './InfoItem';
import { itemIteratorId } from './helpers';

const EmptyInfoItem = ({ template, attr }) => (
  <InfoItem itemId={itemIteratorId(template, attr)} />
);

EmptyInfoItem.propTypes = {
  template: PropTypes.object.isRequired,
  attr: PropTypes.string.isRequired,
};

export default EmptyInfoItem;
