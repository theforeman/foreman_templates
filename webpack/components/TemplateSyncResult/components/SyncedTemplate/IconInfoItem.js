import React from 'react';
import { Icon } from 'patternfly-react';
import PropTypes from 'prop-types';

import InfoItem from './InfoItem';
import { itemIteratorId } from './helpers';

const IconInfoItem = ({ template, attr, iconName, tooltipText }) => (
  <InfoItem itemId={itemIteratorId(template, attr)} tooltipText={tooltipText}>
    <Icon type="fa" name={iconName}/>
  </InfoItem>
);

IconInfoItem.propTypes = {
  template: PropTypes.object.isRequired,
  attr: PropTypes.string.isRequired,
  iconName: PropTypes.string.isRequired,
  tooltipText: PropTypes.string.isRequired,
};

export default IconInfoItem;
