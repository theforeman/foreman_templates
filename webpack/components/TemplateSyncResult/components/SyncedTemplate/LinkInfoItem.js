import React from 'react';
import { ListView } from 'patternfly-react';

import { itemIteratorId } from './helpers';

const LinkInfoItem = ({ template, editPath, attr }) => {
  if (template.id && template.canEdit) {
    return (
      <ListView.InfoItem
        key={itemIteratorId(template, attr)}
        className="additional-info-wide"
      >
        <a
          href={editPath.replace(':id', template.id)}
          target="_blank"
          rel="noopener noreferrer"
        >
          {template.name}
        </a>
      </ListView.InfoItem>
    );
  }
  return template.name || ' ';
};

export default LinkInfoItem;
