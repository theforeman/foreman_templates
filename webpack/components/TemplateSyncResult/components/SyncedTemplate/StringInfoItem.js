import React from 'react';
import EllipsisWithTooltip from 'react-ellipsis-with-tooltip';
import PropTypes from 'prop-types';

import InfoItem from './InfoItem';
import { itemIteratorId } from './helpers';

const StringInfoItem = ({
  template,
  attr,
  tooltipText,
  translate,
  mapAttr,
  elipsed,
}) => {
  const inner = (
    <span>
      {translate ? __(mapAttr(template, attr)) : mapAttr(template, attr)}
    </span>
  );
  const innerContent = elipsed ? (
    <EllipsisWithTooltip placement="top">{inner}</EllipsisWithTooltip>
  ) : (
    inner
  );

  return (
    <InfoItem itemId={itemIteratorId(template, attr)} tooltipText={tooltipText}>
      {innerContent}
    </InfoItem>
  );
};

StringInfoItem.propTypes = {
  template: PropTypes.object.isRequired,
  attr: PropTypes.string.isRequired,
  tooltipText: PropTypes.string,
  translate: PropTypes.bool,
  mapAttr: PropTypes.func,
  elipsed: PropTypes.bool,
};

StringInfoItem.defaultProps = {
  translate: false,
  mapAttr: (template, attr) => template[attr],
  elipsed: false,
  tooltipText: undefined,
};

export default StringInfoItem;
