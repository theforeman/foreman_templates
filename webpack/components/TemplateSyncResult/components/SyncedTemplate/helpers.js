import React from 'react';
import { isEmpty } from 'lodash';
import { Icon } from 'patternfly-react';

import IconInfoItem from './IconInfoItem';
import EmptyInfoItem from './EmptyInfoItem';
import StringInfoItem from './StringInfoItem';

export const itemIteratorId = (template, attr) => `${template.name}-${attr}`;

export const additionalInfo = template => {
  const infoAttrs = [
    'locked',
    'snippet',
    'humanizedClassName',
    'kind',
    'templateFile',
  ];

  return infoAttrs.map(attr => {
    const key = itemIteratorId(template, attr);

    const classNameMap = { Ptable: 'Partition Table' };

    if (!template[attr]) {
      return <EmptyInfoItem template={template} attr={attr} key={key} />;
    }

    switch (attr) {
      case 'locked':
        return (
          <IconInfoItem
            template={template}
            attr={attr}
            iconName="lock"
            tooltipText="Locked"
            key={key}
          />
        );
      case 'snippet':
        return (
          <IconInfoItem
            template={template}
            attr={attr}
            iconName="check"
            tooltipText="Snippet"
            key={key}
          />
        );
      case 'humanizedClassName':
        return (
          <StringInfoItem
            template={template}
            attr={attr}
            translate
            mapAttr={(templateObj, attribute) =>
              classNameMap[templateObj[attribute]]
                ? classNameMap[templateObj[attribute]]
                : templateObj[attribute]
            }
            key={key}
          />
        );
      case 'kind':
        return <StringInfoItem template={template} attr={attr} key={key} />;
      case 'templateFile':
        return (
          <StringInfoItem template={template} attr={attr} key={key} elipsed />
        );
      default:
        return '';
    }
  });
};

export const itemLeftContentIcon = template => {
  const iconName = isEmpty(aggregatedErrors(template))
    ? 'ok'
    : 'error-circle-o';
  return <Icon name={iconName} size="sm" type="pf" />;
};

export const templateHeading = (template, editPath) => {
  if (template.id && template.canEdit) {
    return (
      <a
        href={editPath.replace(':id', template.id)}
        target="_blank"
        rel="noopener noreferrer"
      >
        {template.name}
      </a>
    );
  }
  return template.name || ' ';
};

export const templateErrors = template => {
  if (Object.keys(aggregatedErrors(template)).length !== 0) {
    const res = Object.keys(aggregatedErrors(template)).map(key => (
      <li key={itemIteratorId(template, key)}>
        {formatError(key, aggregatedErrors(template)[key])}
      </li>
    ));
    return <ul>{res}</ul>;
  }
  return <span>There were no errors.</span>;
};

const aggregatedErrors = template => {
  const err = { ...template.errors } || {};
  if (template.additional_errors) {
    err.additional = template.additional_errors;
  }

  return err;
};

const formatError = (key, value) => {
  const omitKeys = ['base', 'additional'];
  if (omitKeys.reduce((memo, item) => memo || key === item, false)) {
    return value;
  }

  return `${key}: ${value}`;
};
