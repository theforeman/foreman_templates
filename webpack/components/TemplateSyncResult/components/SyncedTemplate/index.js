import React from 'react';
import { ListView } from 'patternfly-react';
import PropTypes from 'prop-types';

import { templateErrors, itemLeftContentIcon, additionalInfo } from './helpers';

const SyncedTemplate = ({ template, editPath }) => (
  <ListView.Item
    key={template.id}
    additionalInfo={additionalInfo(template, editPath)}
    className="listViewItem--listItemVariants"
    leftContent={itemLeftContentIcon(template)}
    hideCloseIcon
    stacked
  >
    {templateErrors(template)}
  </ListView.Item>
);

SyncedTemplate.propTypes = {
  template: PropTypes.object.isRequired,
  editPath: PropTypes.string,
};

SyncedTemplate.defaultProps = {
  editPath: '',
};

export default SyncedTemplate;
