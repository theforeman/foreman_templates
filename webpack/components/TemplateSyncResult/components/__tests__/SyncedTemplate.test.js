import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';
import SyncedTemplate from '../SyncedTemplate';
import {
  noName,
  epel,
  coreos,
  filteredOut,
} from '../../__fixtures__/templateSyncResult.fixtures';

const fixtures = {
  'should render template with invalid metadata': {
    template: noName,
    editPath: '',
  },
  'should render template with validation errors': {
    template: epel,
    editPath: '/templates/:id/edit',
  },
  'should render template without errros': {
    template: coreos,
    editPath: '/ptables/:id/edit',
  },
  'should render skipped template': {
    template: filteredOut,
    editPath: '/ptables/:id/edit',
  },
};

describe('SyncedTemplate', () =>
  testComponentSnapshotsWithFixtures(SyncedTemplate, fixtures));
