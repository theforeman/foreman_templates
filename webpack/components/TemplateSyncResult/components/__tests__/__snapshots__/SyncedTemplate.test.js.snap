// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SyncedTemplate should render template with invalid metadata 1`] = `
<ListViewItem
  actions={null}
  additionalInfo={
    Array [
      <EmptyInfoItem
        attr="name"
        template={
          Object {
            "additionalErrors": "No \\"name\\" found in metadata",
            "errors": null,
            "name": null,
            "templateFile": "random_template.erb",
          }
        }
      />,
      <EmptyInfoItem
        attr="locked"
        template={
          Object {
            "additionalErrors": "No \\"name\\" found in metadata",
            "errors": null,
            "name": null,
            "templateFile": "random_template.erb",
          }
        }
      />,
      <EmptyInfoItem
        attr="snippet"
        template={
          Object {
            "additionalErrors": "No \\"name\\" found in metadata",
            "errors": null,
            "name": null,
            "templateFile": "random_template.erb",
          }
        }
      />,
      <EmptyInfoItem
        attr="humanizedClassName"
        template={
          Object {
            "additionalErrors": "No \\"name\\" found in metadata",
            "errors": null,
            "name": null,
            "templateFile": "random_template.erb",
          }
        }
      />,
      <EmptyInfoItem
        attr="kind"
        template={
          Object {
            "additionalErrors": "No \\"name\\" found in metadata",
            "errors": null,
            "name": null,
            "templateFile": "random_template.erb",
          }
        }
      />,
      <StringInfoItem
        attr="templateFile"
        elipsed={true}
        mapAttr={[Function]}
        template={
          Object {
            "additionalErrors": "No \\"name\\" found in metadata",
            "errors": null,
            "name": null,
            "templateFile": "random_template.erb",
          }
        }
        translate={false}
      />,
    ]
  }
  checkboxInput={null}
  className="listViewItem--listItemVariants"
  compoundExpand={false}
  compoundExpanded={false}
  description={null}
  heading={null}
  hideCloseIcon={true}
  initExpanded={false}
  leftContent={
    <Icon
      name="ok"
      size="sm"
      type="pf"
    />
  }
  onCloseCompoundExpand={[Function]}
  onExpand={[Function]}
  onExpandClose={[Function]}
  stacked={true}
>
  <span>
    There were no errors.
  </span>
</ListViewItem>
`;

exports[`SyncedTemplate should render template with validation errors 1`] = `
<ListViewItem
  actions={null}
  additionalInfo={
    Array [
      <LinkInfoItem
        attr="name"
        editPath="/templates/:id/edit"
        template={
          Object {
            "additionalErrors": null,
            "className": "ProvisioningTemplate",
            "errors": Object {
              "base": Array [
                "This template is locked",
              ],
            },
            "humanizedClassName": "Provisioning Template",
            "kind": "Provision",
            "locked": true,
            "name": "EPEL",
            "snippet": true,
            "templateFile": "epel.erb",
          }
        }
      />,
      <IconInfoItem
        attr="locked"
        iconName="lock"
        template={
          Object {
            "additionalErrors": null,
            "className": "ProvisioningTemplate",
            "errors": Object {
              "base": Array [
                "This template is locked",
              ],
            },
            "humanizedClassName": "Provisioning Template",
            "kind": "Provision",
            "locked": true,
            "name": "EPEL",
            "snippet": true,
            "templateFile": "epel.erb",
          }
        }
        tooltipText="Locked"
      />,
      <IconInfoItem
        attr="snippet"
        iconName="check"
        template={
          Object {
            "additionalErrors": null,
            "className": "ProvisioningTemplate",
            "errors": Object {
              "base": Array [
                "This template is locked",
              ],
            },
            "humanizedClassName": "Provisioning Template",
            "kind": "Provision",
            "locked": true,
            "name": "EPEL",
            "snippet": true,
            "templateFile": "epel.erb",
          }
        }
        tooltipText="Snippet"
      />,
      <StringInfoItem
        attr="humanizedClassName"
        elipsed={false}
        mapAttr={[Function]}
        template={
          Object {
            "additionalErrors": null,
            "className": "ProvisioningTemplate",
            "errors": Object {
              "base": Array [
                "This template is locked",
              ],
            },
            "humanizedClassName": "Provisioning Template",
            "kind": "Provision",
            "locked": true,
            "name": "EPEL",
            "snippet": true,
            "templateFile": "epel.erb",
          }
        }
        translate={true}
      />,
      <StringInfoItem
        attr="kind"
        elipsed={false}
        mapAttr={[Function]}
        template={
          Object {
            "additionalErrors": null,
            "className": "ProvisioningTemplate",
            "errors": Object {
              "base": Array [
                "This template is locked",
              ],
            },
            "humanizedClassName": "Provisioning Template",
            "kind": "Provision",
            "locked": true,
            "name": "EPEL",
            "snippet": true,
            "templateFile": "epel.erb",
          }
        }
        translate={false}
      />,
      <StringInfoItem
        attr="templateFile"
        elipsed={true}
        mapAttr={[Function]}
        template={
          Object {
            "additionalErrors": null,
            "className": "ProvisioningTemplate",
            "errors": Object {
              "base": Array [
                "This template is locked",
              ],
            },
            "humanizedClassName": "Provisioning Template",
            "kind": "Provision",
            "locked": true,
            "name": "EPEL",
            "snippet": true,
            "templateFile": "epel.erb",
          }
        }
        translate={false}
      />,
    ]
  }
  checkboxInput={null}
  className="listViewItem--listItemVariants"
  compoundExpand={false}
  compoundExpanded={false}
  description={null}
  heading={null}
  hideCloseIcon={true}
  initExpanded={false}
  leftContent={
    <Icon
      name="error-circle-o"
      size="sm"
      type="pf"
    />
  }
  onCloseCompoundExpand={[Function]}
  onExpand={[Function]}
  onExpandClose={[Function]}
  stacked={true}
>
  <ul>
    <li
      key="EPEL-base"
    >
      This template is locked
    </li>
  </ul>
</ListViewItem>
`;

exports[`SyncedTemplate should render template without errros 1`] = `
<ListViewItem
  actions={null}
  additionalInfo={
    Array [
      <LinkInfoItem
        attr="name"
        editPath="/ptables/:id/edit"
        template={
          Object {
            "additionalErrors": null,
            "className": "Ptable",
            "errors": Object {},
            "humanizedClassName": "Ptable",
            "locked": false,
            "name": "CoreOS default",
            "snippet": false,
            "templateFile": "coreos_default.erb",
          }
        }
      />,
      <EmptyInfoItem
        attr="locked"
        template={
          Object {
            "additionalErrors": null,
            "className": "Ptable",
            "errors": Object {},
            "humanizedClassName": "Ptable",
            "locked": false,
            "name": "CoreOS default",
            "snippet": false,
            "templateFile": "coreos_default.erb",
          }
        }
      />,
      <EmptyInfoItem
        attr="snippet"
        template={
          Object {
            "additionalErrors": null,
            "className": "Ptable",
            "errors": Object {},
            "humanizedClassName": "Ptable",
            "locked": false,
            "name": "CoreOS default",
            "snippet": false,
            "templateFile": "coreos_default.erb",
          }
        }
      />,
      <StringInfoItem
        attr="humanizedClassName"
        elipsed={false}
        mapAttr={[Function]}
        template={
          Object {
            "additionalErrors": null,
            "className": "Ptable",
            "errors": Object {},
            "humanizedClassName": "Ptable",
            "locked": false,
            "name": "CoreOS default",
            "snippet": false,
            "templateFile": "coreos_default.erb",
          }
        }
        translate={true}
      />,
      <EmptyInfoItem
        attr="kind"
        template={
          Object {
            "additionalErrors": null,
            "className": "Ptable",
            "errors": Object {},
            "humanizedClassName": "Ptable",
            "locked": false,
            "name": "CoreOS default",
            "snippet": false,
            "templateFile": "coreos_default.erb",
          }
        }
      />,
      <StringInfoItem
        attr="templateFile"
        elipsed={true}
        mapAttr={[Function]}
        template={
          Object {
            "additionalErrors": null,
            "className": "Ptable",
            "errors": Object {},
            "humanizedClassName": "Ptable",
            "locked": false,
            "name": "CoreOS default",
            "snippet": false,
            "templateFile": "coreos_default.erb",
          }
        }
        translate={false}
      />,
    ]
  }
  checkboxInput={null}
  className="listViewItem--listItemVariants"
  compoundExpand={false}
  compoundExpanded={false}
  description={null}
  heading={null}
  hideCloseIcon={true}
  initExpanded={false}
  leftContent={
    <Icon
      name="ok"
      size="sm"
      type="pf"
    />
  }
  onCloseCompoundExpand={[Function]}
  onExpand={[Function]}
  onExpandClose={[Function]}
  stacked={true}
>
  <span>
    There were no errors.
  </span>
</ListViewItem>
`;
